doc/tutorials/inter-vm_communication.rst

```diff
@@ -0,0 +1,490 @@
+needs to design a data transfer notification mechanism between the VMs.
+
+**The following reference code are used for Inter-VM communication using ivshmem device**
+
+- Initial Device:
+
+  Enable and bind to UIO PCI driver
+    .. code-block:: none
+
+       modprobe uio
+       modprobe uio_pci_generic
+       echo "1af4 1110" > /sys/bus/pci/drivers/uio_pci_generic/new_id
+
+    .. note:: "1af4" is the Vendor ID and "1110" is the Device ID of the ivshmem device.
+
+    .. note:: For Linux-based User VMs, we recommend using the standard UIO and UIO_PCI_GENERIC
+     drivers through the device node (for example, ``/dev/uioX``)
+
+- Uio irq data Struct
+
+     .. code-block:: none
+
+        struct uio_irq_data
+        {
+            int fd;
+            int vector;
+        };
+
+- Ivshmem Device Context Struct:
+
+     .. code-block:: none
+
+        struct ivsh_dev_context
+        {
+            long uio_nr;
+            int bar0_fd;
+            uint32_t *p_reg;
+
+            int bar2_fd;
+            void *p_shmem;
+            long shmem_size;
+
+            /* used for doorbell mode */
+            int uio_dev_fd;
+            int epfds_irq[IVSH_MAX_IRQ_NUM];
+            struct uio_irq_data irq_data[IVSH_MAX_IRQ_NUM];
+            bool opened;
+        };
+
+- Init ivshmem device context
+
+     .. code-block:: none
+
+        int ivsh_init_dev_ctx(struct ivsh_dev_context *p_ivsh_dev_ctx, long uio_nr)
+        {
+            int i;
+            memset(p_ivsh_dev_ctx, 0, sizeof(*p_ivsh_dev_ctx));
+            p_ivsh_dev_ctx->uio_nr = uio_nr;
+            p_ivsh_dev_ctx->bar0_fd = -1;
+            p_ivsh_dev_ctx->bar2_fd = -1;
+            p_ivsh_dev_ctx->uio_dev_fd = -1;
+
+            for (i = 0; i < IVSH_MAX_IRQ_NUM; i++) {
+                p_ivsh_dev_ctx->epfds_irq[i] = -1;
+                p_ivsh_dev_ctx->irq_data[i].fd = -1;
+            }
+            p_ivsh_dev_ctx->opened = false;
+            return 0;
+        }
+
+- Get ivshmem device shared memory size
+
+     .. code-block:: none
+        :emphasize-lines: 10
+
+        uint32_t ivsh_get_shmem_size(long uio_nr)
+        {
+            char config_node[PATH_MAX] = {0};
+            uint32_t shm_size;
+            uint64_t tmp;
+            int cfg_fd;
+
+            sprintf(config_node, "/sys/class/uio/uio%ld/device/config", uio_nr);
+            cfg_fd = open(config_node, O_RDWR);
+            /*Get the size of BAR2(offset in Configuration Space is 0x18) of uio_nr device*/
+            pread(cfg_fd, &tmp, 8, 0x18);
+            shm_size= ~0U;
+            pwrite(cfg_fd ,&shm_size, 8, 0x18);
+            pread(cfg_fd, &shm_size, 8, 0x18);
+            pwrite(cfg_fd ,&tmp, 8, 0x18);
+            shm_size &= (~0xfUL);
+            shm_size = (shm_size & ~(shm_size - 1));
+            close(cfg_fd);
+
+            return shm_size;
+        }
+
+- Open Ivshmem Device:
+
+     .. code-block:: none
+        :emphasize-lines: 1,6,12,18,22
```

Comment:
I'm curious why you're emphasizing some of the comments

Reply:
Thanks, has deleted these emphasize-lines.